crates/broker/src/storage.rs

```diff
@@ -347,18 +348,20 @@ pub async fn upload_image_uri(
     request: &crate::ProofRequest,
     config: &crate::config::ConfigLock,
 ) -> Result<String> {
-    let required_image_id = Digest::from(request.requirements.imageId.0);
-    let image_id_str = required_image_id.to_string();
-    if prover.has_image(&image_id_str).await? {
-        tracing::debug!(
-            "Skipping program upload for cached image ID: {image_id_str} for request {:x}",
-            request.id
-        );
-        return Ok(image_id_str);
+    let image_id_str = request.requirements.image_id().map(|image_id| image_id.to_string());
+    // When predicate is ClaimDigestMatch, we do not have the image id, so we must always download and upload the image.
```

Comment:
This is true and worth thinking if this might create a (rate limit) issue when used often. One solution could be to make sure the requestor knows the image url must guarantee some high frequency access, or we should think to cache based on the url, however that is risky as the content might change. Anyway, just food for thoughts for now